Changing 'rails new' --master to be --main

MikeRogers0 opened the pull request:
### Summary

I thought a bit weird we have a `--master` flag when would track master when we now use `main`.

I updated the flag to be `--main`/`--no-main`, which will create a new project tracking the main branch explicitly in the Gemfile.

### Other Information

This is totally my first pull request to `rails/rails`, so bear with me if I've missed something really obvious.

The only thing I'm thinking we might want is a alias for `--master`/`--no-master` so that tracks the `main` branch also, what do you think?

eileencodes commented:
I think we need to alias this or at least have the previous flag raise an error pointing the user at the new flag. Otherwise if anyone is using this functionality it will end up just raising an undefined error which may be confusing.

MikeRogers0 committed:
Changing 'rails new' --master to be --main

Renaming test containing flag

Updating other test referencing master branch

Add notice that --master is deprecated, but still working the same as --main

Only set @main if it's nil

Making warn wildcard

I think a hidden aliaes would be just as good

Improving description & fixing rubocop error

Forgot comma

Deprecation warning was kind of hard - so just doing alias for now

rubocop -a

MikeRogers0 committed:
I think passing in the --master argument to run_generator is the way to go

MikeRogers0 committed:
Removing .count to figure out why its failing

MikeRogers0 commented:
@eileencodes 👋

I added an alias method, which seemed to work :) Though I wasn't quite sure how to add a nice deprecation method for when it's used, do you have any suggestions on how I can add one?

rafaelfranca commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -861,7 +861,12 @@ def test_edge_option
   end
 
   def test_master_option
```
```suggestion
  def test_main_option
```

ghiculescu replied:
that test is defined separately https://github.com/rails/rails/pull/41141/files#diff-9dd1c2e31fd5617506e344d6328e0e165614290478e1180ef0630017b9305ba4R868

rafaelfranca replied:
Ah, good point.

marivaldo commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -861,7 +861,12 @@ def test_edge_option
 
   def test_master_option
-    generator([destination_root], master: true, skip_webpack_install: true)
+    run_generator [destination_root, "--master"]
+    assert_file "Gemfile", %r{^gem\s+["']rails["'],\s+github:\s+["']#{Regexp.escape("rails/rails")}["'],\s+branch:\s+["']main["']$}
```
```suggestion
    run_generator([destination_root, "--master"])
    assert_file("Gemfile", %r{^gem\s+["']rails["'],\s+github:\s+["']#{Regexp.escape("rails/rails")}["'],\s+branch:\s+["']main["']$})
```

marivaldo commented on railties/test/generators/app_generator_test.rb:
```diff
@@ -809,8 +809,8 @@ def test_web_console_with_edge_option
-  def test_web_console_with_master_option
-    run_generator [destination_root, "--master"]
+  def test_web_console_with_main_option
+    run_generator [destination_root, "--main"]
```
```suggestion
    run_generator([destination_root, "--main"])
```

rafaelfranca replied:
No need to make those changes.

rafaelfranca commented:
> I added an alias method, which seemed to work :) Though I wasn't quite sure how to add a nice deprecation notice for when it's used, do you have any suggestions on how I can add one?

I don't think this option is used enough to warrant a deprecation. It was just introduced in Rails 6.1 and it doesn't always work. Maybe for a next version of Rails we can take care of showing an error instead of just allowing the aliases.
